### ISSUE TITLE
kucoin watch_order_book initial snapshot error

### ISSUE THREAD

**Issue by jaseisen105** [ccxt/ccxt#6820]
- OS: ubuntu 16.04
- Programming Language version: python 3.6.5
- CCXT version: 0.1.67

Happens intermittently when trying to subscribe to many pairs.  Once it happens the stream never recovers.  Would be great if there was a way to detect stale streams and automatically restart

Having trouble reproducing locally, I think it's related to rate limiting where kucoin will return a blank response or something.  Is it possible to add a retry on the initial book snapshot?

enableRateLimiting is set to True here 

```
  File "/python3.6/site-packages/ccxtpro/kucoin.py", line 227, in watch_order_book
    return await self.after(future, self.limit_order_book, symbol, limit, params)
  File "/python3.6/site-packages/ccxtpro/base/exchange.py", line 77, in after
    return method(await future, *args)
  File "/python3.6/site-packages/ccxtpro/base/exchange.py", line 80, in after_async
    return await method(await future, *args)
  File "/python3.6/site-packages/ccxtpro/kucoin.py", line 111, in subscribe
    return await self.watch(url, messageHash, request, messageHash, subscription)
  File "/python3.6/site-packages/ccxtpro/kucoin.py", line 243, in fetch_order_book_snapshot
    previousSequence = sequenceStart - 1
TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'
```

**Comment by kroitor** [comment]
> Is it possible to add a retry on the initial book snapshot?

The retry is already there:

- https://github.com/kroitor/ccxt.pro/blob/master/js/kucoin.js#L256

> TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'
> Happens intermittently when trying to subscribe to many pairs.

Pardon, must be a leftover problem after the bid/ask fix, I will resolve this shortly!

**Comment by kroitor** [comment]
@jaseisen105 can you plz show your list of symbols you're subscribing to?

**Comment by jaseisen105** [comment]
@kroitor - 

Pretty much all of them, here's some that I've seen failed:

```
XDB/USDT
SENSO/USDT
ROAD/USDT
POL/USDT
ACOIN/USDT
XNS/USDT
ZEC/USDT
ATOM/USDT
```

I was thinking maybe it's just limited to low volume pairs, but some of these pairs definitely have activity

**Comment by kroitor** [comment]
@jaseisen105 the problem happened when the orderbook is not too active (has rare delta updates) – that caused receiving the snapshot before the first message, making it impossible to synchronize properly. KuCoin is less reliable in regard of synchronization than other exchanges, because it does not serve a WS snapshot and requires to fetch a REST snapshot instead followed by synchronization with the feed. And if the feed updates are rare, there's nothing to synchronize with, until the first delta arrives. 

Added a fix for it, and tested on my side with the above symbols – so it should work normally as of CCXT Pro version 0.1.68 (the new build will arrive in 15 minutes or so). Please, let me know if you still have issues with it after upgrading. Thx again for your feedback and pardon for the inconvenience.

**Comment by jaseisen105** [comment]
Hi @kroitor -

Thanks for updating.  I'm trying again and am not seeing that error anymore.  Unfortunately am seeing a lot more of the below, assuming it's related to this fix but perhaps not:

```
File "lib/python3.6/site-packages/ccxtpro/kucoin.py", line 227, in watch_order_book
    return await self.after(future, self.limit_order_book, symbol, limit, params)
  File "lib/python3.6/site-packages/ccxtpro/base/exchange.py", line 77, in after
    return method(await future, *args)
  File "lib/python3.6/site-packages/ccxtpro/base/exchange.py", line 80, in after_async
    return await method(await future, *args)
  File "lib/python3.6/site-packages/ccxtpro/kucoin.py", line 111, in subscribe
    return await self.watch(url, messageHash, request, messageHash, subscription)
  File "lib/python3.6/site-packages/ccxtpro/kucoin.py", line 265, in fetch_order_book_snapshot
    raise InvalidNonce(self.id + ' failed to synchronize WebSocket feed with the snapshot for symbol ' + symbol + ' in ' + str(maxAttempts) + ' attempts')
ccxt.base.errors.InvalidNonce: kucoin failed to synchronize WebSocket feed with the snapshot for symbol XRP/ETH in 3 attempts
```

**Comment by kroitor** [comment]
@jaseisen105 can you show your code to reproduce it?

I'm using this:

```Python
# -*- coding: utf-8 -*-

import asyncio
import ccxtpro

async def loop(exchange, symbol):
    while True:
        try:
            orderbook = await exchange.watch_order_book(symbol)
            now = exchange.milliseconds()
            print(exchange.iso8601(now), symbol, orderbook['asks'][0], orderbook['bids'][0])
        except Exception as e:
            print(str(e))
            # raise e  # uncomment to break all loops in case of an error in any one of them
            # break  # you can also break just this one loop if it fails

async def main():
    exchange = ccxtpro.kucoin({'enableRateLimit': True})
    symbols = ['XRP/ETH', 'LTC/ETH']

    await asyncio.gather(*[loop(exchange, symbol) for symbol in symbols])
    await exchange.close()

if __name__ == '__main__':
    asyncio.get_event_loop().run_until_complete(main())
```

And it appears to work...

**Comment by kroitor** [comment]
@jaseisen105 added some more fixes to it in CCXT Pro 0.1.69. Your feedback is really appreciated!

**Comment by jaseisen105** [comment]
Hi @kroitor  - Just updated and tried again but no dice, still getting the same error.  When I try locally as a script I can't seem to reproduce it as well.  It's only in production.  Wondering if it's maybe timing out due to cpu constraints on startup.  I've tried adding a bunch of awaits to make sure all pairs don't subscribe at once.  Doesn't seem to help though.

**Comment by kroitor** [comment]
@jaseisen105 can you share more details on your production config, versions, code to reproduce, etc? Does the [snippet above](https://github.com/ccxt/ccxt/issues/6820#issuecomment-616225072) work on your production set up with the most recent version of the lib?

**Comment by jaseisen105** [comment]
@kroitor yes that snippet works.  I can't reproduce it in a single script and unfortunately can't share production code easily as it's quite large.

ccxt version: 0.1.69
python 3.6.5
ubuntu 16.04

Regardless of this issue, it would be great if these streams could be more resilient to failures.  Right now, after retrying 3 times it gives up and never retries again.  That means every time I start my app I have to hope for no failures because it will never recover.

Instead it would be great if we could retry continuously like on an exponential backoff or similar.  I'm ok with getting these errors frequently on startup, as long as it is able to retry and recover within a decent time frame.

**Comment by kroitor** [comment]
@jaseisen105 ok, we have some ideas on how to add more robustness to it, will let you know! Thx again!

**Comment by jaseisen105** [comment]
Great, thank you!

**Comment by kroitor** [comment]
@jaseisen105 in the meantime, you can mitigate it by overriding the max number of attempts via your options:

```Python
exchange = ccxtpro.kucoin({
    'enableRateLimit': True,
    'fetchOrderBookSnapshot': { 'maxAttempts': 1000 },
})
```

**Comment by kroitor** [comment]
P.S. @jaseisen105 it would help a lot to resolve this quicker if you could reduce the code down to a reproducible snippet (since it's easier to address if we know exactly which problem to address)

**Comment by kroitor** [comment]
@jaseisen105 alternatively, your verbose output might also help:

```Python
exchange = ccxtpro.kucoin({
    'enableRateLimit': True,
    'verbose': True,  # ←---------------- to enable verbose mode
})

...
```

↑ That will be a lot of output, if you're listening for multiple streams, but if the error pops quickly – you can paste the output on gist.github.com or upload it here as a text file.

**Comment by kroitor** [comment]
@jaseisen105 

> It's happening on pairs like XRP/ETH, LTC/ETH where volume is somewhat low but not non-existent

Pardon, one more quick question – does it still happen with the same pairs? If not, can you list them out? (make sure your version is 0.1.69+).

**Comment by jaseisen105** [comment]
> @jaseisen105 in the meantime, you can try to mitigate it by overriding the max number of attempts via your options:
> 
> ```python
> exchange = ccxtpro.kucoin({
>     'enableRateLimit': True,
>     'fetchOrderBookSnapshot': { 'maxAttempts': 1000 },
> })
> ```

Tried this, but didn't seem to have any affect.  Still retries 3 times



> @jaseisen105 alternatively, your verbose output might also help:
> 
> ```python
> exchange = ccxtpro.kucoin({
>     'enableRateLimit': True,
>     'verbose': True,  # ←---------------- to enable verbose mode
> })
> 
> ...
> ```
> 
> ↑ That will be a lot of output, if you're listening for multiple streams, but if the error pops up quickly – you can paste the output on gist.github.com or upload it here as a text file.

Was able to get a log snippet, there might be sensitive data in there so I'll email it to you.



> @jaseisen105
> 
> > It's happening on pairs like XRP/ETH, LTC/ETH where volume is somewhat low but not non-existent
> 
> Pardon, one more quick question – does it still happen with the same pairs? If not, can you list them out? (make sure your version is 0.1.69+).

The pairs seem to be random to me, not noticing particular ones having issues or a pattern e.g. only small volume pairs.  I've seen it happen to larger pairs too.

**Comment by kroitor** [comment]
> Tried this, but didn't seem to have any affect. Still retries 3 times

This is weird, since it should override the number of attempts. Do you create the exchange instance once and reuse it throughout the code?

**Comment by kroitor** [comment]
Oh, sorry, i made a mistake in the example above, it should be inside `'options'`:

```Python
exchange = ccxtpro.kucoin({
    'enableRateLimit': True,
    'options': {
        'fetchOrderBookSnapshot': { 'maxAttempts': 1000 },
    },
})
```

↑ This way it will retry 1000 times.

**Comment by kroitor** [comment]
@jaseisen105 we have added one more solution that should also mitigate the problem (it's already available in 0.1.69, so you have it already), you can add a warmup delay to make sure that the snapshot sequence number is older than the first delta:

```
exchange = ccxtpro.kucoin({
    'enableRateLimit': True,
    'options': {
        'fetchOrderBookSnapshot': {
            'maxAttempts': 1000,  # max attempts to fetch the snapshot
            'delay': 1000,  # a warmup delay to wait before attempting to fetch the snapshot
        },
    },
})
```

↑ Adding a warmup delay should also help increase the reliability of synchronization with KuCoin.
Let me know if that helps or not.

(got the email with the logs, the problem is that KuCoin returns a snapshot too old, compared to the WS delta feed, probably due to caching on the webservers – so, an added delay should help). Please, try values from 1000-5000 (1-5 seconds) for the `'delay'` and let me know if that helps or not. 

Thx!

**Comment by kroitor** [comment]
@jaseisen105 and another quick question – do you have any caching http/https proxies on the production system?

**Comment by jaseisen105** [comment]
> @jaseisen105 we have added one more solution that should also mitigate the problem (it's already available in 0.1.69, so you have it already), you can add a warmup delay to make sure that the snapshot sequence number is older than the first delta:
> 
> ```python
> exchange = ccxtpro.kucoin({
>     'enableRateLimit': True,
>     'options': {
>         'fetchOrderBookSnapshot': {
>             'maxAttempts': 1000,  # max attempts to fetch the snapshot
>             'delay': 1000,  # a warmup delay in milliseconds before attempting to fetch the snapshot
>         },
>     },
> })
> ```
> 
> ↑ Adding a warmup delay should also help increase the reliability of synchronization with KuCoin.
> Let me know if that helps or not.
> 
> (I got the email with the logs from you, public feeds do not include sensitive info). The problem is that KuCoin returns a snapshot too old, compared to the WS delta feed, probably due to caching on the webservers – so, an added delay should help). Please, try values from 1000-5000 (1-5 seconds) for the `'delay'` and let me know if that helps or not.
> 
> Thx!

Looks like this works, awesome thanks!



> @jaseisen105 and another quick question – do you have any caching http/https proxies on the production system? (just to rule them out).

None

**Comment by kroitor** [comment]
> Looks like this works, awesome thanks!

Yep, basically, you can play with those settings to adjust for the networking environment. Let me know if you experience any further difficulties with it. In the meantime we will try address that in a more robust way. Thx for your involvement, appreciate it!
